### DISCUSSION TITLE
Is it noticably more efficient to cache direct-memory arrays such as `ts.nodes_time`?

### DISCUSSION THREAD

**Discussion by hyanwong** [tskit-dev/tskit#3064] [Q&A] [answered]
I'm quite often looking up node times in a tree sequence, which can be accessed using `ts.nodes_time`. If I'm e.g. accessing it 4 or 5 times, am I better saving it as a variable, or is `ts.nodes_time` almost as efficient:

```python
parent_times = ts.nodes_time[ts.edges_parent]
child_times = ts.nodes_time[ts.edges_child]

# Or is this a lot more efficient?
nodes_time = ts.nodes_time
parent_times = nodes_time[ts.edges_parent]
child_times = nodes_time[ts.edges_child]
```

**Comment by hyanwong** [comment]
`ts.nodes_time` provides zero copy access, but there are a few function calls involved, including an attribute accessor, so it is likely that the second incantation will be more efficient, but not by much, unless it's inside a tight loop.

**Reply by jeromekelleher** [reply, marked as answer]
There wouldn't be much in it here. 

It does make a substantial difference when accessing table columns, which superficially looks the same but involves full copies.
